File: utilities/benchmark.py
"""Comprehensive benchmarking module for LLM models."""

import json
import platform
from dataclasses import asdict, dataclass
from datetime import datetime
from pathlib import Path
from typing import List

import yaml
from common import CONFIG_PATH, PERF_ROOT
from run_bench import bench_once_llama


@dataclass
class BenchmarkConfig:
    """Configuration for benchmark runs."""

    prompt_set: str
    prompt: str
    max_tokens: int = 100
    temperature: float = 0.7
    iterations: int = 3
    host: str = "localhost"
    port: int = 8080


@dataclass
class BenchmarkIteration:
    """Single benchmark iteration result."""

    wall_s: float
    output_text: str
    prompt_tokens: int
    generated_tokens: int
    tok_per_s: float
    generation_tok_per_s: float
    ttft_ms: float
    prefill_ms: float
    generation_ms: float


@dataclass
class SystemInfo:
    """System information for benchmarks."""

    platform: str
    architecture: str
    processor: str

    @classmethod
    def get_current_system(cls):
        return cls(
            platform=platform.system(),
            architecture=platform.machine(),
            processor=platform.processor() or platform.machine(),
        )


@dataclass
class BenchmarkSummary:
    """Summary statistics for benchmark results."""

    median_wall_s: float
    median_tok_per_s: float
    median_ttft_ms: float
    median_generation_tok_per_s: float


@dataclass
class ModelBenchmarkResult:
    """Complete benchmark result for a model."""

    timestamp: str
    repo_id: str
    model_ref: str
    engine: str
    system_info: SystemInfo
    config: BenchmarkConfig
    iterations: List[BenchmarkIteration]
    summary: BenchmarkSummary
    mode: str = "text-only"
    environment: str = "stable"


class BenchmarkRunner:
    """Runs comprehensive benchmarks on LLM models."""

    def __init__(self):
        self.results_dir = PERF_ROOT
        self.results_dir.mkdir(parents=True, exist_ok=True)

    def run_multiple_benchmarks(
        self,
        repo_id: str,
        model_path: str,
        engine: str = "llama-server",
        configs: List[BenchmarkConfig] = [],
    ) -> List[ModelBenchmarkResult]:
        """Run multiple benchmark configurations."""
        if not configs:
            return []

        results = []
        for config in configs:
            try:
                result = self.run_benchmark(repo_id, model_path, config, engine)
                results.append(result)
            except Exception:
                continue
        return results

    def run_benchmark(
        self,
        repo_id: str,
        model_path: str,
        config: BenchmarkConfig,
        engine: str = "llama-server",  # Default to llama-server
    ) -> ModelBenchmarkResult:
        """Run complete benchmark for a model."""

        iterations = []

        for _ in range(config.iterations):
            try:
                result = bench_once_llama(
                    prompt=config.prompt,
                    max_tokens=config.max_tokens,
                    temperature=config.temperature,
                    host=config.host,
                    port=config.port,
                )

                iteration = BenchmarkIteration(
                    wall_s=result["wall_s"],
                    output_text=result["output_text"],
                    prompt_tokens=result["prompt_tokens"],
                    generated_tokens=result["generated_tokens"],
                    tok_per_s=result["tok_per_s"],
                    generation_tok_per_s=result["generation_tok_per_s"],
                    ttft_ms=result["ttft_ms"],
                    prefill_ms=result["ttft_ms"],
                    generation_ms=result["generation_ms"],
                )

                iterations.append(iteration)
            except Exception:
                continue

        if not iterations:
            raise RuntimeError("All benchmark iterations failed")

        # Calculate summary statistics
        summary = self._calculate_summary(iterations)

        # Create complete result
        benchmark_result = ModelBenchmarkResult(
            timestamp=datetime.now().isoformat(),
            repo_id=repo_id,
            model_ref=model_path,
            engine=engine,
            system_info=SystemInfo.get_current_system(),
            config=config,
            iterations=iterations,
            summary=summary,
        )

        return benchmark_result

    def _calculate_summary(
        self, iterations: List[BenchmarkIteration]
    ) -> BenchmarkSummary:
        """Calculate summary statistics from iterations."""
        wall_times = sorted([i.wall_s for i in iterations])
        tok_per_s_values = sorted([i.tok_per_s for i in iterations])
        ttft_values = sorted([i.ttft_ms for i in iterations])
        gen_tok_per_s_values = sorted([i.generation_tok_per_s for i in iterations])

        n = len(iterations)
        median_idx = n // 2

        return BenchmarkSummary(
            median_wall_s=wall_times[median_idx],
            median_tok_per_s=tok_per_s_values[median_idx],
            median_ttft_ms=ttft_values[median_idx],
            median_generation_tok_per_s=gen_tok_per_s_values[median_idx],
        )

    def save_results(self, result: ModelBenchmarkResult) -> Path:
        """Save benchmark results to JSON file with proper naming."""
        # Create filename: perf/2025-12-09_repo_name_engine_mode.json
        timestamp = datetime.fromisoformat(result.timestamp).strftime("%Y-%m-%d")

        # Clean repo_id for filename
        repo_clean = result.repo_id.replace("/", "_").replace(".", "_")
        engine_clean = result.engine.replace("-", "_")

        filename = f"{timestamp}_{repo_clean}_{engine_clean}_{result.mode}.json"
        file_path = self.results_dir / filename

        # Convert to dict for JSON serialization
        data = asdict(result)

        with open(file_path, "w") as f:
            json.dump(data, f, indent=2)

        return file_path

    def print_summary(self, result: ModelBenchmarkResult):
        """Print benchmark summary to console."""
        print("\n" + "=" * 60)
        print(f"📊 BENCHMARK RESULTS: {result.repo_id}")
        print("=" * 60)
        print(f"Model: {result.model_ref}")
        print(f"Engine: {result.engine}")
        print(f"Timestamp: {result.timestamp}")
        print(f"Iterations: {len(result.iterations)}")

        print("\nPerformance Metrics (Median):")
        print(f"  Tokens/sec:     {result.summary.median_tok_per_s:.2f}")
        print(f"  TTFT (ms):      {result.summary.median_ttft_ms:.1f}")
        print(f"  Wall Time:      {result.summary.median_wall_s:.2f}s")
        print(f"  Gen Tokens/sec: {result.summary.median_generation_tok_per_s:.2f}")

        print("\nSystem Info:")
        print(f"  Platform:       {result.system_info.platform}")
        print(f"  Architecture:   {result.system_info.architecture}")
        print(f"  Processor:      {result.system_info.processor}")
        print("=" * 60)


def load_config():
    """Load full config from YAML file."""
    try:
        with open(CONFIG_PATH, "r") as f:
            return yaml.safe_load(f)
    except Exception:
        return {}


def load_benchmark_questions():
    """Load benchmark questions from config file."""
    config = load_config()
    return config.get("benchmark_questions", {}).get("hard_questions", [])


def load_benchmark_config(benchmark_name: str) -> BenchmarkConfig:
    """Load a specific benchmark configuration from config file."""
    config = load_config()
    benchmark_data = config.get(benchmark_name, {})

    if not benchmark_data:
        raise ValueError(f"Benchmark '{benchmark_name}' not found in config")

    return BenchmarkConfig(
        prompt_set=benchmark_data.get("prompt_set", benchmark_name),
        prompt=benchmark_data["prompt"],
        max_tokens=benchmark_data.get("max_tokens", 100),
        temperature=benchmark_data.get("temperature", 0.7),
        iterations=benchmark_data.get("iterations", 3),
        host=benchmark_data.get("host", "localhost"),
        port=benchmark_data.get("port", 8080),
    )


def create_hard_questions_benchmark() -> List[BenchmarkConfig]:
    """Create benchmark configs from hard questions in config file."""
    questions = load_benchmark_questions()
    benchmarks = []

    for i, q in enumerate(questions, 1):
        config = BenchmarkConfig(
            prompt_set=f"hard_question_{i}",
            prompt=q["prompt"],
            max_tokens=q["max_tokens"],
            temperature=0.7,
            iterations=3,
        )
        benchmarks.append(config)

    return benchmarks


# Load benchmark configurations from config file
QUICK_BENCHMARK = load_benchmark_config("quick_benchmark")
STANDARD_BENCHMARK = load_benchmark_config("standard_benchmark")
PERFORMANCE_BENCHMARK = load_benchmark_config("performance_benchmark")
HARD_QUESTIONS_BENCHMARK = create_hard_questions_benchmark()

File: src/benchmark.cpp
#include "benchmark.h"
#include <nlohmann/json.hpp>
#include <algorithm>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <sys/sysctl.h>
#include <sys/utsname.h>
#include <curl/curl.h>

// ── helpers ────────────────────────────────────────────────────────────────

static std::string now_iso8601() {
    auto now   = std::chrono::system_clock::now();
    auto now_t = std::chrono::system_clock::to_time_t(now);
    auto us    = std::chrono::duration_cast<std::chrono::microseconds>(
                     now.time_since_epoch()) % 1'000'000;
    std::ostringstream oss;
    oss << std::put_time(std::localtime(&now_t), "%Y-%m-%dT%H:%M:%S");
    oss << "." << std::setfill('0') << std::setw(6) << us.count();
    return oss.str();
}

static double vec_median(std::vector<double> v) {
    std::sort(v.begin(), v.end());
    return v[v.size() / 2];
}

static size_t write_cb(void* ptr, size_t size, size_t nmemb, std::string* s) {
    s->append(static_cast<char*>(ptr), size * nmemb);
    return size * nmemb;
}

static std::string http_post_json(const std::string& url, const std::string& body) {
    CURL* curl = curl_easy_init();
    if (!curl) throw std::runtime_error("curl_easy_init failed");

    std::string response;
    curl_slist* headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");

    curl_easy_setopt(curl, CURLOPT_URL,           url.c_str());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS,    body.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER,    headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA,     &response);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT,       300L);

    CURLcode res = curl_easy_perform(curl);
    long code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &code);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK)
        throw std::runtime_error(std::string("curl: ") + curl_easy_strerror(res));
    if (code != 200)
        throw std::runtime_error("HTTP " + std::to_string(code));
    return response;
}

// ── SystemInfo ─────────────────────────────────────────────────────────────

SystemInfo get_system_info() {
    SystemInfo info;
    struct utsname u{};
    uname(&u);
    info.platform     = u.sysname;
    info.architecture = u.machine;

    char buf[256]{};
    size_t sz = sizeof(buf);
    if (sysctlbyname("machdep.cpu.brand_string", buf, &sz, nullptr, 0) == 0)
        info.processor = buf;
    else
        info.processor = info.architecture;

    return info;
}

// ── single inference call ──────────────────────────────────────────────────

BenchmarkIteration run_once(const BenchmarkConfig& cfg) {
    std::string url = "http://" + cfg.host + ":" + std::to_string(cfg.port) + "/completion";

    nlohmann::json payload = {
        {"prompt",      cfg.prompt},
        {"n_predict",   cfg.max_tokens},
        {"temperature", cfg.temperature},
    };

    auto t0       = std::chrono::steady_clock::now();
    auto body     = http_post_json(url, payload.dump());
    auto t1       = std::chrono::steady_clock::now();
    double wall_s = std::chrono::duration<double>(t1 - t0).count();

    auto j       = nlohmann::json::parse(body);
    auto timings = j.value("timings", nlohmann::json::object());

    BenchmarkIteration it;
    it.wall_s               = wall_s;
    it.output_text          = j.value("content", "");
    it.prompt_tokens        = j.value("tokens_evaluated", 0);
    it.ttft_ms              = timings.value("prompt_ms", 0.0);
    it.prefill_ms           = it.ttft_ms;
    it.generated_tokens     = timings.value("predicted_n", 0);
    it.generation_ms        = timings.value("predicted_ms", 0.0);
    it.generation_tok_per_s = timings.value("predicted_per_second", 0.0);
    it.tok_per_s            = wall_s > 0 ? it.generated_tokens / wall_s : 0.0;

    return it;
}

// ── full benchmark run ─────────────────────────────────────────────────────

BenchmarkResult run_benchmark(const std::string& model_path, const BenchmarkConfig& cfg) {
    std::vector<BenchmarkIteration> iters;

    for (int i = 0; i < cfg.iterations; i++) {
        try {
            auto it = run_once(cfg);
            std::cout << "  [" << (i + 1) << "/" << cfg.iterations << "] "
                      << std::fixed << std::setprecision(1)
                      << it.tok_per_s << " tok/s  "
                      << it.generated_tokens << " tokens  "
                      << it.wall_s << "s\n";
            iters.push_back(it);
        } catch (const std::exception& e) {
            std::cerr << "  [" << (i + 1) << "/" << cfg.iterations
                      << "] failed: " << e.what() << "\n";
        }
    }

    if (iters.empty())
        throw std::runtime_error("All benchmark iterations failed");

    std::vector<double> walls, toks, ttfts, gen_toks;
    for (auto& it : iters) {
        walls.push_back(it.wall_s);
        toks.push_back(it.tok_per_s);
        ttfts.push_back(it.ttft_ms);
        gen_toks.push_back(it.generation_tok_per_s);
    }

    BenchmarkSummary summary;
    summary.median_wall_s               = vec_median(walls);
    summary.median_tok_per_s            = vec_median(toks);
    summary.median_ttft_ms              = vec_median(ttfts);
    summary.median_generation_tok_per_s = vec_median(gen_toks);

    std::string stem    = std::filesystem::path(model_path).stem().string();
    std::string repo_id = "local/" + stem;

    BenchmarkResult result;
    result.timestamp   = now_iso8601();
    result.repo_id     = repo_id;
    result.model_ref   = model_path;
    result.system_info = get_system_info();
    result.config      = cfg;
    result.iterations  = iters;
    result.summary     = summary;

    return result;
}

// ── JSON output ────────────────────────────────────────────────────────────

void save_result(const BenchmarkResult& r, const std::filesystem::path& perf_dir) {
    std::filesystem::create_directories(perf_dir);

    std::string date = r.timestamp.substr(0, 10);

    auto clean = [](std::string s, char from, char to) {
        for (char& c : s) if (c == from) c = to;
        return s;
    };
    std::string repo_clean   = clean(clean(r.repo_id, '/', '_'), '.', '_');
    std::string engine_clean = clean(r.engine, '-', '_');
    std::string filename     = date + "_" + repo_clean + "_" + engine_clean + "_" + r.mode + ".json";

    nlohmann::json j;
    j["timestamp"]   = r.timestamp;
    j["repo_id"]     = r.repo_id;
    j["model_ref"]   = r.model_ref;
    j["engine"]      = r.engine;
    j["system_info"] = {
        {"platform",     r.system_info.platform},
        {"architecture", r.system_info.architecture},
        {"processor",    r.system_info.processor},
    };
    j["config"] = {
        {"prompt_set",  r.config.prompt_set},
        {"prompt",      r.config.prompt},
        {"max_tokens",  r.config.max_tokens},
        {"temperature", r.config.temperature},
        {"iterations",  r.config.iterations},
        {"host",        r.config.host},
        {"port",        r.config.port},
    };

    auto iters = nlohmann::json::array();
    for (auto& it : r.iterations) {
        iters.push_back({
            {"wall_s",               it.wall_s},
            {"output_text",          it.output_text},
            {"prompt_tokens",        it.prompt_tokens},
            {"generated_tokens",     it.generated_tokens},
            {"tok_per_s",            it.tok_per_s},
            {"generation_tok_per_s", it.generation_tok_per_s},
            {"ttft_ms",              it.ttft_ms},
            {"prefill_ms",           it.prefill_ms},
            {"generation_ms",        it.generation_ms},
        });
    }
    j["iterations"] = iters;
    j["summary"] = {
        {"median_wall_s",               r.summary.median_wall_s},
        {"median_tok_per_s",            r.summary.median_tok_per_s},
        {"median_ttft_ms",              r.summary.median_ttft_ms},
        {"median_generation_tok_per_s", r.summary.median_generation_tok_per_s},
    };
    j["mode"]        = r.mode;
    j["environment"] = r.environment;

    auto path = perf_dir / filename;
    std::ofstream f(path);
    f << j.dump(2) << "\n";

    std::cout << "[kepler] Results saved: " << path.string() << "\n";
}

// ── console summary ────────────────────────────────────────────────────────

void print_summary(const BenchmarkResult& r) {
    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "BENCHMARK RESULTS: " << r.repo_id << "\n";
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Model:      " << r.model_ref << "\n";
    std::cout << "Preset:     " << r.config.prompt_set << "\n";
    std::cout << "Iterations: " << r.iterations.size() << "\n";
    std::cout << std::fixed;
    std::cout << "\nMedian performance:\n";
    std::cout << "  tok/s (overall):  " << std::setprecision(2) << r.summary.median_tok_per_s << "\n";
    std::cout << "  tok/s (gen only): " << std::setprecision(2) << r.summary.median_generation_tok_per_s << "\n";
    std::cout << "  TTFT:             " << std::setprecision(1) << r.summary.median_ttft_ms << " ms\n";
    std::cout << "  Wall time:        " << std::setprecision(2) << r.summary.median_wall_s << " s\n";
    std::cout << "\nSystem: " << r.system_info.platform << " "
              << r.system_info.architecture << " — " << r.system_info.processor << "\n";
    std::cout << std::string(60, '=') << "\n";
}
